schema/scope.go

```diff
@@ -123,6 +124,38 @@ func (s *ScopeSchema) Objects() map[string]*ObjectSchema {
 	return s.ObjectsValue
 }
 
+func (s *ScopeSchema) objectIDList() string {
+	output := ""
+	for id := range s.ObjectsValue {
+		output += "\n" + id
+	}
+	return output
+}
+
+func (s *ScopeSchema) RootObject() *ObjectSchema {
+	rootObject, rootObjectFound := s.ObjectsValue[s.RootValue]
+	if !rootObjectFound {
+		panic(fmt.Sprintf(
+			"root object with ID %q not found; available objects:%s",
+			s.RootValue,
+			s.objectIDList(),
+		))
+	}
+	if rootObject == nil {
+		panic(fmt.Sprintf(
+			"root object with ID %q is nil; this is a bug",
+			s.RootValue,
+		))
+	}
+	if rootObject.ID() != s.RootValue {
+		panic(fmt.Sprintf(
+			"root object's ID %q doesn't match its map key %q; please fix the input",
+			rootObject.ID(), s.RootValue,
+		))
+	}
```

Comment:
We should not be panicking as a result of bad, user-supplied input.  (There are several reasons for this, but the obvious one is that a user is going to see the "panic" and traceback and assume that it's not their error.)

I realize that it is cumbersome to deal with this error in any other way, here.  So, I recommend leaving this code substantially as it (except I suggest removing the phrase "please fix the input") and adding a similar check at an appropriate point upstream in the code where we _can_ report an error instead of panicking.

Reply:
> The current design isn't setup for this. It would require a pretty big re-write to allow errors.

I concur (as I said in my initial comment) that returning an error _here_ would be problematic.  However, I'm unwilling to believe that there is no point upstream in the code where we can easily stop and return an error (although, I'll grant, it might be tricky to detect this problem at that point).

> if we do fix this later by adding an earlier check, that doesn't change that this is likely a user error in the input

If we add an earlier check and the check here still triggers, then that is **not** a user error -- that's an implementation bug:  the behavior _should_ be that we return an error for a user error and panic only for internal errors.

> I think we should keep that for now.

I have no problem keeping the check here...but, it should not be possible for a user to trigger it (once we have the code properly detecting and handling the user error).

> I think we should keep this for now.

By "this", I'm guessing (pronouns are evil! 👿) that you mean the message, "please fix the input".  I submit that that message doesn't have enough traction to obtain the result that you're hoping for.  If you want to keep it, I would up it to something more like, "the specified root object ID, %q, doesn't match its map key, %q -- this is a user error -- please correct the input specification in the workflow YAML file".

Reply:
Continued [below](https://github.com/arcalot/arcaflow-plugin-sdk-go/pull/85#discussion_r1572729483).

And tracked in https://github.com/arcalot/arcaflow-plugin-sdk-go/issues/86.